### ISSUE TITLE
Support FileProvider API (VFS) in macOS

### ISSUE THREAD

**Issue by helmut72** [nextcloud/desktop#1337]
https://www.apple.com/macos/catalina-preview/features/

> **Third-party cloud service integration**
> An all-new FileProvider API for cloud storage providers delivers a new way to seamlessly integrate their services into the Finder without requiring a kernel extension, helping to maintain the security of your Mac. Cloud storage providers can now deliver their apps through the Mac App Store.

**Comment by pgauret** [comment]
Now that Catalina is out may be a good opportunity to look into FileProvider. This would allow to modernise the macOS desktop client and bring equivalent functionality to the virtual drive for Windows client.

**Comment by theJian** [comment]
Catalina is out but I can't find a way to create FileProviderExtension on macOS. Does anyone notice that FileProvider is actually removed from [Catalina features page](https://www.apple.com/macos/catalina/features/)?

**Comment by mikelynch** [comment]
It's strange that it's not listed as a feature in the release, but the Developer documentation does say it's available [in macOS 10.15+](https://developer.apple.com/documentation/fileprovider). I'm afraid I don't know anything about macOS development, but there are some _First Steps_ topics there. @theJian, do those provide any useful guidance for getting started?

It's also just occured to me that this is already implemented [in the iOS app](https://github.com/nextcloud/ios/tree/master/File%20Provider%20Extension), and hopefully that component should also work on macOS. Would that give an easier route to using this API?

One consideration is that presumably the current logic will have to be kept for the moment, in order to support older versions of macOS – I imagine it wouldn't be desirable to require Catalina for a while yet. It'd be great to be support this API as an option though!

**Comment by theJian** [comment]
@mikelynch  I didn't find those useful. Luckily once you created a FileProviderExtension, Xcode will generate template code for you. It's a working simple demo and full of comments. Getting started by reading those code is much easy in my opinion.

**Comment by pointlessone** [comment]
Documentation is a little inconsistent. On the [File Provider Framework](https://developer.apple.com/documentation/fileprovider) page it states availability to be macOS 10.15+ but on the individual protocol pages (e.g. [NSFileProviderReplicatedExtension](https://developer.apple.com/documentation/fileprovider/nsfileproviderreplicatedextension)) it's macOS 11.0+. I guess this feature didn't make it into Catalina but since Big Sur is around the corner maybe it's time to revisit this issue?

**Comment by leiless** [comment]
FYI, [File Provider - macOS Support](https://developer.apple.com/documentation/fileprovider/macos_support-ari?language=objc) seems available by now.

**Comment by SlavDimov** [comment]
Is there a similar API for Linux?

**Comment by aniqueta** [comment]
Is this something being worked on? With Big Sur changes, it seems safe to assume that in the next macOS version using the FileProvider API will become mandatory and the Finder extension approach will be deprecated. Thanks.

**Comment by leiless** [comment]
> Is this something being worked on?

Don't know.

It seems that `FileProvider` not ready yet, and Apple hasn't given any sample code on it.

**Comment by hemancuso** [comment]
@leiless we actually just shipped Nextcloud support in our FileProvider product (Strongsync). Definitely is ready.

**Comment by leiless** [comment]
> @leiless we actually just shipped Nextcloud support in our FileProvider product ([Strongsync](https://www.expandrive.com/strongsync)). Definitely is ready. Happy to answer questions.

Does there any sample code on `FileProvider` API, I've checked Apple Developer docs, haven't seeing it.

**Comment by hemancuso** [comment]
I believe the very latest XCode betas generate a template project

**Comment by helmut72** [comment]
> @leiless we actually just shipped Nextcloud support in our FileProvider product ([Strongsync](https://www.expandrive.com/strongsync)). Definitely is ready. Happy to answer questions.

Target: common Apache2 WebDAV

Tasks:
1. Created `test folder` on macOS: ok, also on the server
2. Copied my Notes (a lot of Markdown files in `private` and `work` with subfolders): failed. `Work` and all subfolders are on the server, no subfolder in `Private`. On both no single file was uploaded!
3. Deleted `test folder`: still on the server (empty trashcan on macOS doesn't helped)
4. Result: uninstalled Strongsync

**Comment by andreasley** [comment]
Xcode 12.5, which was released today, contains a File Provider Extension target template to help getting started with macOS File Provider extensions.

**Comment by leiless** [comment]
> Xcode 12.5, which was released today, contains a File Provider Extension target template to help getting started with macOS File Provider extensions.

Thanks! FYI: https://developer.apple.com/documentation/xcode-release-notes/xcode-12_5-release-notes

**Comment by aniqueta** [comment]
Glad to see the activity here. Looks like the API is maturing in macOS Monterey. FYI: Box has adopted it now: https://blog.box.com/get-more-secure-seamless-box-drive-experience-across-macs

**Comment by acsfer** [comment]
See https://github.com/nextcloud/desktop/blob/master/doc/architecture.rst#virtual-files for how to enable experimental support on v3.4.0.

**Comment by pszypowicz** [comment]
> See https://github.com/nextcloud/desktop/blob/master/doc/architecture.rst#virtual-files for how to enable experimental support on v3.4.0.

Hi, is this coded with a FileProviderAPI? Could you point to a commit/PR implementing this?

**Comment by aniqueta** [comment]
Note that OneDrive [1] and Dropbox [2] are stating there are changes coming in macOS 12.3 (March?) that is making them fully migrate over to the FileProvider API, with macOS 12.2 being the last version they support with their current virtual files / online files implementation. I can't find anything from Apple indicating what these changes are, but I suspect we'll find out soon... 

[1] https://techcommunity.microsoft.com/t5/microsoft-onedrive-blog/inside-the-new-files-on-demand-experience-on-macos/bc-p/3066588
[2] https://help.dropbox.com/installs-integrations/desktop/macos-12-monterey-support

**Comment by jeffmancuso** [comment]
They both have exceptions to the no kernel extension policy allowing them to use their previous solutions. I imagine this exception is finally going away in a late-cycle point release rather than in macOS 13. This increases their pressure to actually migrate over to FPE.

**Comment by marcotrevisan** [comment]
Just for the records: https://developer.apple.com/documentation/fileprovider/macos_support/syncing_files_on_macos.

Would be great to have the system take care of most of the synchronization work for you, and you only have to focus on the client calls and the extra features.

**Comment by alex0411** [comment]
Hi, 
are there any updates when this feature will be pulled into the development / released? At the moment this missing feature is holding is back to use Nextcloud in a productive environment, since a lot of our users are using macos based clients and using the experimental option in the client is not an option for a productive environment. 

Thanks a lot!

**Comment by alex0411** [comment]
> Hi, are there any updates when this feature will be pulled into the development / released? At the moment this missing feature is holding us back to use Nextcloud in a productive environment, since a lot of our users are using macos based clients and using the experimental option in the client is not an option for a productive environment.
> 
> Thanks a lot!

Push

**Comment by gaodahai** [comment]
> Hi, are there any updates when this feature will be pulled into the development / released? At the moment this missing feature is holding us back to use Nextcloud in a productive environment, since a lot of our users are using macos based clients and using the experimental option in the client is not an option for a productive environment.
> 
> Thanks a lot!

Hi, we also just saw the mentioned troubles happen. Is there any possibility to get this fixed? Or how do you work to avoid these troubles?
Thank you very much!

**Comment by alex0411** [comment]
> Hi, are there any updates when this feature will be pulled into the development / released? At the moment this missing feature is holding us back to use Nextcloud in a productive environment, since a lot of our users are using macos based clients and using the experimental option in the client is not an option for a productive environment.
> 
> Thanks a lot!

Push again. Maybe somebody could give us some feedback here.

**Comment by marcotrevisan** [comment]
I'm not a NC developer but seems like the team is working on improving the current sync engine quality, even if I was unable to properly check for progress in the FileProvider support (anyway NextcloudKit was created and received merges). 
#4970, #5016, #5011 are just examples of that.
I'm just guessing but the dev team may be currently preferring focusing on the "SyncEngine", making it better.
In this case I'd urge implementing #4464 which in my humble opinion would make it work hugely better on all platforms.
I can live with ".nextcloud" files, much less with a sync engine that scans the whole world.
Thanks!

**Comment by hemancuso** [comment]
Only posting again on this thread, just in case people are curious or want some help on an implementation. I had implemented support in a product I make back in 2021 [comment](https://github.com/nextcloud/desktop/issues/1337#issuecomment-824178037) and will be soon adding support to ExpanDrive. Whoever on the team ends up taking on this project feel free to ping me if you need assistance.

**Comment by alex0411** [comment]
Hi, is there any feedback already available? I can just share my oberservation within our use-case: Our mac users are refusing to use nextcloud because of this missing functionalities and to stay on the file storage software. So it would be really import to get any insights about this topics and the priorities behind it.
Thanks a lot!

**Comment by mamboone** [comment]
> Hi, is there any feedback already available? I can just share my oberservation within our use-case: Our mac users are refusing to use nextcloud because of this missing functionalities and to stay on the file storage software. So it would be really import to get any insights about this topics and the priorities behind it.
> Thanks a lot!

I can just add to this comment. Eagerly waiting for Nextcloud to support the "on demand" feature right away. Looking on the OneDrive side from Microsoft - there this feature works quite fine for months already with Mac.

**Comment by peterthomashorn** [comment]
Just to throw in my personal experience "two-cents" in working with file providers for the people reading along like me: It is not that simple and might be more complex to wire up to the Nextcloud client code and API (which I am not familiar with). This are not simple adaptions but quite different processes compared to classic synchronization. Implementations from iOS cannot be simply ported to macOS due to the API differences.

**Comment by tobiasKaminsky** [comment]
We are working on it, but as @peterthomashorn outlined, it is really not that simple, unfortunately.
As soon as we have an alpha/beta version, we will inform you here and ask for feedback.

**Comment by marcotrevisan** [comment]
I'd eagerly alpha/betatest it. Consider me added to the list if you like. 
Also, if it can be of any help I'd vote for a bare-bones Files-only client implementation, no Talk and no other stuff, just focused on Files operation (including locks and push).

**Comment by alex0411** [comment]
Hi, as feedback after several months without going forward with this feature. We decided to sunset the productive usage of Nextcloud and will use another platform.

**Comment by marcotrevisan** [comment]
Just for the records, since one of their latest updates the Dropbox client for mac now shows up as a shared position and the icons for the sync state are pretty much the same as iCloud’s.
Looks like it’s now a FileProvider based implementation, even if it was one of the snappiest clients out there… which imho further justifies the efforts for this work ;)

**Comment by claucambra** [comment]
This is currently top of my to-do list and is being actively worked on. There's no timeline or anything like that, and it's not even close to done at the moment, but we have made some nice progress to get the File Provider app extension to communicate with the desktop client. Lots left to do still 🙂

**Comment by piermorel** [comment]
For those following what Dropbox and others are doing with File Provider, it seems that it always leads to major loss of functionality : it breaks spotlight, files **have** to be on the main drive (this one is mind boggling when you see the cost of storage with Apple), and the functionality to always keep a folder local is not very trustworthy and its slower (see [dropbox documentation](https://help.dropbox.com/en-en/installs/macos-support-for-expected-changes)).

So shouldn't the focus be to find solutions to keep on supporting macOS **without** using file provider until Apple fixes it?

**Comment by claucambra** [comment]
> For those following what Dropbox and others are doing with File Provider, it seems that it always leads to major loss of functionality : it breaks spotlight, files **have** to be on the main drive (this one is mind boggling when you see the cost of storage with Apple), and the functionality to always keep a folder local is not very trustworthy and its slower (see [dropbox documentation](https://help.dropbox.com/en-en/installs/macos-support-for-expected-changes)).
> 
> So shouldn't the focus be to find solutions to keep on supporting macOS **without** using file provider until Apple fixes it?

We do not have any plans to remove conventional file synchronisation from the desktop client -- like VFS, FileProvider will be an option

**Comment by claucambra** [comment]
Quick progress report -- file browsing is now working! 🙂


https://user-images.githubusercontent.com/70155116/216428719-3f6601df-66d0-4eea-bc09-f841ab375310.mov

**Comment by jesusdeveloper** [comment]
> Quick progress report -- file browsing is now working! 🙂
> 
>  Screen.Recording.2023-02-02.at.20.16.24.mov

Oh my god! I really wanted to see this because I've been waiting for it for a long time!!!!
Very good work! Hopefully soon we can use it!

Congratulations, good job, keep it up!!!

**Comment by mrtn223** [comment]
Really happy to hear!!!
This is the only reason our company didn't migrate from Dropbox to Nextcloud yet...

**Comment by claucambra** [comment]
File content fetching now working


https://user-images.githubusercontent.com/70155116/219470752-1118b138-56bb-4b1d-b9f6-47454a740db4.mov

**Comment by xavi-b** [comment]
Amazing ! Do you have an estimate on when it would be available in a release ?

**Comment by kallegrens** [comment]
The very question an open source developer doesn't want to hear. ^

**Comment by claucambra** [comment]
> Amazing ! Do you have an estimate on when it would be available in a release ?

When it's finished ;)

**Comment by miguelarios** [comment]
Take your time! We appreciate the great job you are doing! You can't rush greatness. El feb. 17, 2023, a la(s) 6:35 a.m., Claudio Cambra ***@***.***> escribió:﻿

Amazing ! Do you have an estimate on when it would be available in a release ?

When it's finished ;)

**Comment by footlooseboss** [comment]
> File content fetching now working
> 

have you looked into partial content fetching? not many apps seem to be making use of it, given how useful it is

[NSFileProviderPartialContentFetching](https://developer.apple.com/documentation/fileprovider/nsfileproviderpartialcontentfetching?changes=latest_minor)

> Support for fetching part of a file’s content.

> Adopt this protocol to let the system request only part of a file. Apps that read files provided by your extension can benefit from this feature, either by minimizing the amount of data your file provider needs to download, or by finishing the download quickly, freeing up the reading process.

> > To trigger a partial download, an app must use POSIX read operations to read part of the file. If you clone the entire file, or read the file using file coordination, the system requests the entire file.

> For example, a photo app could read just the metadata from each picture in a large album, without having to completely download all the images. Alternatively, a video streaming app could begin playing the video before reading the whole file, reading chunks of data just before it needs them.

**Comment by marcotrevisan** [comment]
Since we're in the works, I've just tested the usage of the current client (3.7.3) with the newly released Temporary Files Lock v26.0.0 app enabled.

It seems to be finally working properly in my expected use cases. 

To make it easy living with locks on the client it would be great to have a "lock" icon appearing along with the file in Finder (if I'm not mistaken the FileProvider framework makes it possible to add icons and other graphic decorations to files based on provider's custom logic), because the current way of showing it is not-so-user-friendly (you have to right-click, go to Nextcloud menu item and check if the file is locked). 
The everyday user would easily get frustrated.
Thanks!

**Comment by claucambra** [comment]
Pull request is now open. More details to come...

**Comment by marcotrevisan** [comment]
I'm sure this will be a huge leap forward for Nextcloud Files on Mac OS. 
Hopefully it will also help decreasing PROPFIND load on server.  
Thanks @claucambra and all the dev team!

**Comment by claucambra** [comment]
As those of you who watched the NC26 announcement will know, File Provider support for our desktop client is now a thing!

This has entailed writing a new, macOS-specific sync engine built from the ground up around Apple’s APIs, designed to make the best use of the system’s frameworks.

We’re super excited to release this new engine as part of a new **Nextcloud macOS VFS Beta** application. This app is separate to the standard client. It includes all the characteristics you know and love from the desktop client, with the new virtual files/File Provider module included. 

However, there is a reason we are releasing this separately and why we’re labelling it a beta. Coming from the person who worked in this - file sync is hard. This beta release will have bugs, and while we have not encountered any major data loss, it is not impossible this might happen. So let me be emphatic — **we do not yet recommend the use of the file provider beta app on production instances.**

With warnings out of the way, I have a favour to ask — I would really, really like feedback on any bugs, errors, and crashes that you encounter during use of the beta. The file provider module uses Apple’s unified logging system, so you can extract relevant logs using the Console app or by using the `log` utility from the terminal. To make it easier to find the Nextcloud File Provider related logs, filter by process `FileProviderExt` and by subsystem `com.nextcloud.cloud.FileProviderExt`; if you're using the Console app, make sure to check "Include info messages" and "Include debug messages".

![Screenshot 2023-03-21 at 16 55 06](https://user-images.githubusercontent.com/70155116/226667924-dba0e4ca-1df0-42a7-bcf9-1218833752bc.png)

![Screenshot 2023-03-21 at 16 58 07](https://user-images.githubusercontent.com/70155116/226667967-1628dc8c-f3e7-4c97-b9c4-81002cee2660.png)

When reporting bugs, please make sure to send us this data as well as the conventional debug archive from the desktop client. Also, the more detailed you are in explaining how the problem arose, the better I will be able to go ahead and fix it :) 

A couple of last minute tips:
- For those comfortable using the terminal, the `fileproviderctl` tool is extremely helpful in providing information on each file provider domain. It can also be used to run diagnostics and basic commands on them. Sending any information provided by this tool will be very helpful
- If the Nextcloud macOS VFS beta does not appear in the finder sidebar after setting up your account in the client, restart the desktop client as well as finder, and then it should show up
- Stability is not the only thing left to work on regarding file provider. We have a bunch of file provider related features we plan to start working on that will make the user experience even better!

With that out of the way, have fun (and stay safe!) playing with our beta! You can go ahead and download it from the link below:

https://download.nextcloud.com/desktop/prereleases/virtualdrive/Nextcloud-VFS-beta-macOS.pkg

**Comment by marcotrevisan** [comment]
Great news! @claucambra I've just installed it and I'm trying to make it sync with my existing Nextcloud server.
Do we need to enable some flag in Library/Preferences/Nextcloud macOS VFS Beta/nextcloud-macos-vfs-beta.cfg?
If I enable the experimental options, it seems to enable the old sync, not the FIleProvider based one.

Sorry if I'm asking a silly question.

Thanks again!

**Comment by claucambra** [comment]
> Great news! @claucambra I've just installed it and I'm trying to make it sync with my existing Nextcloud server. Do we need to enable some flag in Library/Preferences/Nextcloud macOS VFS Beta/nextcloud-macos-vfs-beta.cfg? If I enable the experimental options, it seems to enable the old sync, not the FIleProvider based one.
> 
> Sorry if I'm asking a silly question.
> 
> Thanks again!

No hidden config settings need to be enabled, the file provider module should be enabled automatically when your account is configured. Is the Nextcloud extension enabled in the system settings?

If it is still not showing up, there is this point in the big post:

>  If the Nextcloud macOS VFS beta does not appear in the finder sidebar after setting up your account in the client, restart the desktop client as well as finder, and then it should show up

**Comment by marcotrevisan** [comment]
My first feedback.  Before testing I uninstalled the Release NC Client app from the systems under test, just in case. 

On a Big Sur 11.7 - iMac - Intel 2014 it's currently working like a charm. 
I mean: I've set up the account info; specifying that I'm not syncing any subfolder under the root so the legacy sync won't do anything in my home subfolder; closed the account setup and the File Provider share appears under my "Locations" in Finder. 
I can immediately start browsing there and do whatever I want, it just works! It's very responsive.
I have to check if I can configure Finder to [download and] open the preview on space-bar pressed, like it does for files in local folders, but that's another matter - not sure if it's possibile...

On my mac book air (M1, 2020) running MacOS Monterey 12.6.3, after a few clicks opening the root subfolders and some 2nd-level folders, the share won't open subfolders anymore, gets stuck and won't recover (even after reboots). That initial "grace period" has lasted about 30 seconds or so. 
It may be related to your "First sync will take a while" point, but under BigSur things are going quite better.
My colleague did the same under MacOS Monterey 12.4 and same hardware, had the same behaviour. After an initial all-good behaviour, something happens and FInder gets stuck on that share.

On server side, I can see PROPFIND requests arriving in what seems to be a recursive "tree visit", and the user agent is FileProviderExt as I would expect. This holds for both BigSur and Monterey clients. I guess this is how Finder works.

Also, I further aggravated my situation by adding another account (to my home NC 25.0.4 server): Finder got completely stuck, `fileproviderctl listproviders` was also stuck. After waiting for a good 30 minutes I had to do a force shutdown.

I'll let you know more. I'm also gonna try the client on Ventura hosts too very soon!

**Comment by marcotrevisan** [comment]
Sorry I didn't point out that the share is quite populated (>250K files), I guess this is a factor.

**Comment by claucambra** [comment]
> My first feedback. Before testing I uninstalled the Release NC Client app from the systems under test, just in case.
> 
> On a Big Sur 11.7 - iMac - Intel 2014 it's currently working like a charm. I mean: I've set up the account info; specifying that I'm not syncing any subfolder under the root so the legacy sync won't do anything in my home subfolder; closed the account setup and the File Provider share appears under my "Locations" in Finder. I can immediately start browsing there and do whatever I want, it just works! It's very responsive. I have to check if I can configure Finder to [download and] open the preview on space-bar pressed, like it does for files in local folders, but that's another matter - not sure if it's possibile...

Nice!

> On my mac book air (M1, 2020) running MacOS Monterey 12.6.3, after a few clicks opening the root subfolders and some 2nd-level folders, the share won't open subfolders anymore, gets stuck and won't recover (even after reboots). That initial "grace period" has lasted about 30 seconds or so. It may be related to your "First sync will take a while" point, but under BigSur things are going quite better. My colleague did the same under MacOS Monterey 12.4 and same hardware, had the same behaviour. After an initial all-good behaviour, something happens and FInder gets stuck on that share.

The recursive scan begins on first remote change detection, and I wasn't kidding when I said it takes a while -- on my personal instance it takes ~15 minutes, but I have no shares on there. Depending on your network connection and the number of files/folders, it can take a lot longer, and the File Provider extension will not do much until the scan is complete

> On server side, I can see PROPFIND requests arriving in what seems to be a recursive "tree visit", and the user agent is FileProviderExt as I would expect. This holds for both BigSur and Monterey clients. I guess this is how Finder works.

Yup, we are following Apple's advice here.

> Also, I further aggravated my situation by adding another account (to my home NC 25.0.4 server): Finder got completely stuck, `fileproviderctl listproviders` was also stuck. After waiting for a good 30 minutes I had to do a force shutdown.

That's weird; fileproviderctl should be independent from the extension, and each file provider account should work independently of each other...

> I'll let you know more. I'm also gonna try the client on Ventura hosts too very soon!

Cool!

**Comment by marcotrevisan** [comment]
From a end-user perception it's like in BigSur there're more threads doing their job independently (one of them is serving Finder's UI interactions), while in Monterey and Ventura they seem to be synchronized on something, the result being some kind of bottleneck...

**Comment by marcotrevisan** [comment]
I just went back to the BigSur host. 
While navigating trough the share is very responsive, I tried to create a second level folder and the result was the parent folder becoming inaccessible for permission reasons (folder icon with the "red spot"). I guess this is due to the background scanning in progress (so perhaps Monterey and Ventura are keeping stuff frozen to avoid the user getting unwanted results in write operations).
However, if I move to another share in Finder and then go back to the NC location, the parent folder becomes accessible again and I can find the newly created folder inside, with the default "unnamed" name (localized in my language). I can then rename it: the parent folder becomes inaccessible, same trick to get back, and it shows up renamed. The renaming is visible via web in NC Files.

**Comment by marcotrevisan** [comment]
Update. The big work share is now operational on the macbook air M1 with Monterey 12.6.3.
Seems to work pretty smoothly. I tested:

- Basic "cruds"
- Push notifications seem to be catched by the client because I can see a very quick response to add/delete of files in an open folder, and I also can see the PROPFIND requests which are clearly following the path from the root to the modified file.
- Compressing a sub-folder, it works correctly (materializes items, compresses, puts the zip file in the share). That is done using the minimal amount of requests to the server that are needed. Very nice! 
- Copy and paste of a folder - works correctly (materializes items, copy them to target, put them remotely)
- Delete the copied folder (after it's been uploaded) - all good. Found some 404's on PROPFINDs issued on the deleted contents (just guessing: may it be due to a response to notify_push about my own delete?) 
- Looked at the overall PROPFIND load generated by the FileProviderExt: after the first run, and with no activity from other users, I can see substantially no requests issued by the client. Seems very promising!  Other client implementations that we are using behave differently and keep issuing PROPFINDS apparently for nothing.

Tomorrow I'll be able to use it under server load and see how it will behave in that scenario. I'll keep you updated.

**Comment by Kyshman** [comment]
@marcotrevisan am on Monterey on an Intel Mac.  Didn't have any trouble setting up and having VFS appear on the sidebar. It's also quite responsive after the initial scan. But am wondering whether you noticed something though -- All the files/directories have a modification date of 1st Jan 1970 at 03:00

![image](https://user-images.githubusercontent.com/42900317/227146201-d712fd6f-4212-4d73-94dd-43f1850735cf.png)

This is not reflected on the server though where they all maintain their correct modification dates.

Even newly created files have this date too.
![image](https://user-images.githubusercontent.com/42900317/227146450-7663b41c-3d87-485a-ba88-91e3a5321e38.png)

**Comment by marcotrevisan** [comment]
@Kyshman yes the modification times are shown as 1970-01-01. 
If I remember correctly there is no direct retriveal of the modification time, so the current implementation may behave differently.

My problems arised on a quite populated and frequently updated share. On a smaller one I hadn't got any major issues so far.

**Comment by marcotrevisan** [comment]
Update to the above after the sync settled:

1. PROPFINDS calmed down and were limited to the notify_push response pattern;
2. Creating/deleting stuff using the Files on the browser now is reflected in Finder as expected;
3. (no change);
4. It works now. Also, the files formerly having the "exclamation mark cloud" icon have been uploaded and now are in sync;
5.( no change).

I think we need to find a way to get rid of those long-running syncs, or, to allow for some degree of concurrent local changes. Would it be possible to (if this at all makes any sense to you):
- restrict the lookups to the folders the user has actually browsed, and:
- have a (maybe configurable) time to live after which folders are "invalidated" (and thus their contents forgotten)?

Thank you very much!

**Comment by claucambra** [comment]
> Update to the above after the sync settled (took a couple hours):
> 
> 1. PROPFINDS calmed down and were limited to the notify_push response pattern;
> 2. Creating/deleting stuff using the Files on the browser now is reflected in Finder as expected;
> 3. (no change);
> 4. It works now. Also, the files formerly having the "exclamation mark cloud" icon have been uploaded and now are in sync;
>    5.( no change).
> 
> I think we need to find a way to get rid of those long-running syncs, or, to allow for some degree of concurrent local changes. Would it be possible to (if this at all makes any sense to you):
> 
> * restrict the lookups to the folders the user has actually browsed, and:

We tried this initially and caused a lot of issues, especially when it came to A) moving unsynced/unexplored folders, and B) it stops us from being able to detect moves from one explored folder to another, unexplored folder. So you could easily run into a situation where you'd move a big, locally materialised file from folderA to unexploredFolderB, we'd report the file as deleted, and you'd be forced to re-download it. So, not ideal.

Apple's documentation also says we should report the whole server.

> * have a (maybe configurable) time to live after which folders are "invalidated" (and thus their contents forgotten)?

Something like this (particularly manual unmaterialisation of materialised files) is on the to-do list.

> Thank you very much!

Glad to hear no major blow ups have happened yet. But please be careful when running this on your instances, particularly when it comes to concurrent modification of files -- this is still a bit unpredictable

**Comment by marcotrevisan** [comment]
> > I think we need to find a way to get rid of those long-running syncs, or, to allow for some degree of concurrent local changes. Would it be possible to (if this at all makes any sense to you):
> > 
> > * restrict the lookups to the folders the user has actually browsed, and:
> 
> I implemented this initially and caused some of issues, especially when it came to A) moving unsynced/unexplored folders, and B) it stops us from being able to detect moves from one explored folder to another, unexplored folder. So you could easily run into a situation where you'd move a big, locally materialised file from folderA to unexploredFolderB, it'd report the file as deleted, and you'd be forced to re-download it. So, not ideal
> 
> Apple's documentation also says we should report the whole server.

Case B) can be scary sometimes, but I think it would normally pass unnoticed, as teams usually work on the same sets of folders. I'm not sure how many additional cases are actually hiding behind case A)... 
But if Apple recommends to report the whole server I can well understand your decision. 

Lookups apart, what about allowing for more concurrency on local updates? (like allowing two "worker queues", if this makes any sense for you..)?
The FileProviderExt is not flooding the server with PROPFINDs even when it's resyncing, so that's not a problem.
The current problem in my humble opinion is that before that resync has finished, Finder is not behaving consistently.

I would prefer the shortcomings of the above mentioned cases A) and B) over long periods of inconistent behaviour on the folders I usually work on...

Thanks again and regards!

**Comment by claucambra** [comment]
> Lookups apart, what about allowing for more concurrency on local updates? (like allowing two "worker queues", if this makes any sense for you..)? The FileProviderExt is not flooding the server with PROPFINDs even when it's resyncing, so that's not a problem. The current problem in my humble opinion is that before that resync has finished, Finder is not behaving consistently.

We process received data concurrently with a specific concurrent dispatch queue, hence why once we have the data from the server the processing is very quick. With network requests we send them one after the other so as to not overload the server; we could send more at once, within limits, but this will not necessarily yield much of a speed up. Especially if several clients are connected at once. This can be explored more.

With regards to Finder being unresponsive, not much we can do about this at the moment. Experimenting with the extension it seems that Finder will always wait for the current file exploration to be finished before proceeding.

> I would prefer the shortcomings of the above mentioned cases A) and B) over long periods of inconistent behaviour on the folders I usually work on...

This can always be a setting in the future, with the setting favouring correctness the default.

My recommendation for quicker change detection is to favour deep folder hierarchies over wide folder hierarchies. We do a recursive propfind whenever a folder's ETAG changes -- the etag of a folder changes whenever any of its contents change, or whenever any of its children's content changes. So with deep hierarchies we can do quicker scans of smaller folders and focus only on those that have changed. Obviously it's a lot quicker to scan a folder with 30 children than one with 3000 children. 

Hope this helps

**Comment by marcotrevisan** [comment]
> We process received data concurrently with a specific concurrent dispatch queue, hence why once we have the data from the server the processing is very quick. With network requests we send them one after the other so as to not overload the server; we could send more at once, within limits, but this will not necessarily yield much of a speed up. Especially if several clients are connected at once. This can be explored more.

I think we have a few cases worth looking at, i.e. the "create new folder and put a file in that new folder" issue leading to 404's - if we fix it, that may fix more stuff I wasn't able to spot yet.
By the way, I also tested updating some unmaterialized files during the sync process, and the materialization works. I was able to edit a .docx file with no issues and saw the "put" in the server upon save.

> > I would prefer the shortcomings of the above mentioned cases A) and B) over long periods of inconistent behaviour on the folders I usually work on...
> 
> This can always be a setting in the future, with the setting favouring correctness the default.

I'd really like to have that setting available, I'd use it eagerly. After all, G Drive and Dbox in their FileProvider clients must be doing some similar tricks too.. we have another > 2TB data in there and I can assure you no deep sync is performed, or we would be dead... 

> My recommendation for quicker change detection is to favour deep folder hierarchies over wide folder hierarchies. We do a recursive propfind whenever a folder's ETAG changes -- the etag of a folder changes whenever any of its contents change, or whenever any of its children's content changes. So with deep hierarchies we can do quicker scans of smaller folders and focus only on those that have changed. Obviously it's a lot quicker to scan a folder with 30 children than one with 3000 children.

Looks like we already are in a "vertically deep" scenario. Ususally folders don't contain more than 30 items, but they often contain more than 10 subfolders. I guess we're one of the worst cases out there, lol
 
> Hope this helps

Absolutely! Thank you very much for your work! I hope our feedback is helpful too.

**Comment by marcotrevisan** [comment]
I was trying to make use of `fileproviderctl` and I'm monitoring the situation using the `ls` command now.
I noticed something weird when I deleted a test folder (namely "test_marco_2") that I created under the root. 
In Finder it worked apparently without problems, and also remotely it's been removed, everything looks good. 
The `fileproviderctl ls` output however is showing some apparently spurious stuff that might interest you @claucambra , here is a snippet:


[...]
   19: 📁 test_marco_2 id:__fp/fs/fileID(66120888) size:0 ⬆  (indeterminate) (fp error) 🖥 (no url)  oggi, 12:34 (used)
 🏁  collection 🥑  gathering completed in 0.028s
        
collection 🥑 : received updates:

♻️  19: 📁 test_marco_2 id:__fp/fs/fileID(66120888) size:0 ⬆  (indeterminate) (fp error) 🖥 (no url)  oggi, 12:34 (used)


collection 🥑 : received updates:

♻️  19: 📁 test_marco_2 id:__fp/fs/fileID(66120888) size:0 🖥 (no url)  oggi, 12:34 (used)


collection 🥑 : received updates:

♻️  19: 📁 test_marco_2 id:__fp/fs/fileID(66120888) size:0 ⬆  (indeterminate) (fp error) 🖥 (no url)  oggi, 12:34 (used)


collection 🥑 : received updates:

♻️  19: 📁 test_marco_2 id:__fp/fs/fileID(66120888) size:0 🖥 (no url)  oggi, 12:34 (used)


collection 🥑 : received updates:

♻️  19: 📁 test_marco_2 id:__fp/fs/fileID(66120888) size:0 ⬆  (indeterminate) (fp error) 🖥 (no url)  oggi, 12:34 (used)


collection 🥑 : received updates:

♻️  19: 📁 test_marco_2 id:__fp/fs/fileID(66120888) size:0 🖥 (no url)  oggi, 12:34 (used)


collection 🥑 : received updates:

♻️  19: 📁 test_marco_2 id:__fp/fs/fileID(66120888) size:0 ⬆  (indeterminate) (fp error) 🖥 (no url)  oggi, 12:34 (used)


collection 🥑 : received updates:

♻️  19: 📁 test_marco_2 id:__fp/fs/fileID(66120888) size:0 🖥 (no url)  oggi, 12:34 (used)


collection 🥑 : received updates:

♻️  19: 📁 test_marco_2 id:__fp/fs/fileID(66120888) size:0 ⬆  (indeterminate) (fp error) 🖥 (no url)  oggi, 12:34 (used)

collection 🥑 : received updates:

♻️  19: 📁 test_marco_2 id:__fp/fs/fileID(66120888) size:0 🖥 (no url)  oggi, 12:34 (used)


collection 🥑 : received updates:

♻️  19: 📁 test_marco_2 id:__fp/fs/fileID(66120888) size:0 ⬆  (indeterminate) (fp error) 🖥 (no url)  oggi, 12:34 (used)

**Comment by marcotrevisan** [comment]
@strozzascotte I'm testing it against 25.0.4/5 servers (with notify_push app enabled). 
Keep in mind it's a beta so you have to accept a higher risk using it compared to a release version.

**Comment by marcotrevisan** [comment]
Just for the records, some features that I noticed to be missing in Finder:

- Thumbnails on the icons of unmaterialized files, I see in iCloud they're available (they also show up at space-bar-pressed, not great resolution but enough to recognize the file contents), I guess the API allows for that;
-When I right-click on a folder containing unmaterialized stuff, the info panel is not showing the total size (a "--" is shown instead).

About Big Sur: the mount is showing the following issues on write operations:
- Wen a file / folder is created / updated / deleted, the first level folder containing it will suddenly close and show up as "access denied" (i.e. the folder icon with the red symbol on it). After a secondo or so, it will become accessible again. 
   - If you were creating a subfolder, you'll find it with the default "unnamed folder" name, and when you rename it, the same thing will happen (the rename will be successful though). 
   - If you open and edit a document, hitting the Save button will save it, but Finder will do the same thing on the root folder. Not tried extensively yet.
 
So Mac OS 11 Big Sur perhaps needs specific tests. I'll try on another Big Sur host to see if it shares that same behaviour.

OS 13 Ventura: on one machine (MBA 2020 M1) it seems to work exactly the same like mine under OS 12, on another (MBA 2020 M1) it seems stuck (I mean it doesn't even seem to contact the server). We'll try to delete and recreate the mount. Is there a best practice to do that? Looks like I'm unable to delete a FileProvider mount...

Thanks!

**Comment by claucambra** [comment]
> OS 13 Ventura: on one machine (MBA 2020 M1) it seems to work exactly the same like mine under OS 12, on another (MBA 2020 M1) it seems stuck (I mean it doesn't even seem to contact the server). We'll try to delete and recreate the mount. Is there a best practice to do that? Looks like I'm unable to delete a FileProvider mount...


Open terminal and do `fileproviderctl domain remove <name of domain>`

Each of the entries in the sidebar is a domain. You can see which domains are available by running `fileproviderctl domain list`

**Comment by marcotrevisan** [comment]
Big Sur logs, the issue on the firsst level folder becoming unaccessible upon file materialization. Hoping they can be of help:

```
debug	12:22:51.620259+0100	fileproviderd	[DEBUG] ┏4e9a Item at <private> losing presenter <private>
debug	12:22:51.620357+0100	fileproviderd	[DEBUG] ┗4e9a
debug	12:22:51.621203+0100	FileProviderExt	[DEBUG] Attempting to restore persona in -[FPXEnumerator _invalidate]: (null)
debug	12:22:51.621316+0100	FileProviderExt	(504) dropping voucher due to nil personaUniqueString
debug	12:22:51.621397+0100	FileProviderExt	[DEBUG] ┏2e7 <private>: invalidating vendor enumeration: <private>
debug	12:22:51.621471+0100	FileProviderExt	Enumerator is being invalidated for item with identifier: 0074594472f651ffb41de
debug	12:22:51.620773+0100	fileproviderd	[DEBUG] ┣4e9d dispatching to <private>
debug	12:22:51.621562+0100	FileProviderExt	[DEBUG] ┗2e7
debug	12:22:51.621632+0100	FileProviderExt	[DEBUG] Attempting to restore persona in -[FPXEnumerator _invalidate]: (null)
debug	12:22:51.621071+0100	fileproviderd	[DEBUG] ┳4e9d continuing on <private>
debug	12:22:51.621716+0100	FileProviderExt	(504) dropping voucher due to nil personaUniqueString
debug	12:22:51.621196+0100	fileproviderd	[DEBUG] ┣4e9f dispatching to <private>
debug	12:22:51.621830+0100	FileProviderExt	[DEBUG] Attempting to restore persona in -[FPXEnumerator _invalidate]: (null)
debug	12:22:51.621336+0100	fileproviderd	[DEBUG] ┗4e9d
debug	12:22:51.622037+0100	FileProviderExt	(504) dropping voucher due to nil personaUniqueString
debug	12:22:51.621611+0100	fileproviderd	[DEBUG] ┳4e9f continuing on <private>
debug	12:22:51.622477+0100	FileProviderExt	[DEBUG] Attempting to restore persona in -[FPXEnumerator dealloc]: (null)
debug	12:22:51.621794+0100	fileproviderd	[DEBUG] ┗4e9f
debug	12:22:51.622659+0100	FileProviderExt	(504) dropping voucher due to nil personaUniqueString
debug	12:22:51.622884+0100	FileProviderExt	[DEBUG] <private>: dealloc

```

**Comment by TomGem** [comment]
First, thanks a lot! It's great to to see Nextcloud virtual files for macOS coming! 🙏

The beta setup on a Ventura Silicon Mac was fine. Although at first I didn't get I have to check the location 'Nextcloud macOS VFS Beta', not the folder I configured during the setup.. 8)

It took a while to initialise but now the VFS response times are quite ok. 
Btw, the server is NC25 with notify_push enabled.

My first question (it won't be the last :), the modification times are set to SkyNet' Big Bang. 
I've read about this above. However, is it a known issue or a misconfiguration on my side?

<img width="1168" alt="Screenshot 2023-03-26 at 01 30 19" src="https://user-images.githubusercontent.com/8806191/227749002-e0f9c04d-0487-41cc-ad2e-4582e30c6809.png">

**Comment by strozzascotte** [comment]
Thank you very much for this great feature.
I've noticed that if you configure an account on the desktop client, delete it, and configure another account from the same Nextcloud server, the label of the virtual drive doesn't get updated right away.
In the attached image you can see that on the left panel of Finder the selected item is 
 `Nextcloud macOS VFS Beta - admin@...` when in fact it is connected to another account as correctly reported in the status bar of the right panel.
Restarting Finder resolve the issue.
It's definitely just a minor issue, but I hope it's worth reporting it.

<img width="948" alt="Captura de pantalla 2023-03-26 a las 16 19 00" src="https://user-images.githubusercontent.com/3976799/227808673-bd528e9e-1928-4174-b321-b135800c203a.png">

**Comment by marcotrevisan** [comment]
About my above comment: https://github.com/nextcloud/desktop/issues/1337#issuecomment-1482743744

I'm seeing this still happening. 
The folder is among the enumerated ones and the file provider is still trying to sync on that.

**Comment by claucambra** [comment]
> Thank you very much for this great feature. I've noticed that if you configure an account on the desktop client, delete it, and configure another account from the same Nextcloud server, the label of the virtual drive doesn't get updated right away. In the attached image you can see that on the left panel of Finder the selected item is `Nextcloud macOS VFS Beta - admin@...` when in fact it is connected to another account as correctly reported in the status bar of the right panel. Restarting Finder resolve the issue. It's definitely just a minor issue, but I hope it's worth reporting it.

Unfortunately not much we can do about this as this is handled entirely by Finder

**Comment by claucambra** [comment]
> Big Sur logs, the issue on the firsst level folder becoming unaccessible upon file materialization. Hoping they can be of help:
> 
> ```
> debug	12:22:51.620259+0100	fileproviderd	[DEBUG] ┏4e9a Item at <private> losing presenter <private>
> debug	12:22:51.620357+0100	fileproviderd	[DEBUG] ┗4e9a
> debug	12:22:51.621203+0100	FileProviderExt	[DEBUG] Attempting to restore persona in -[FPXEnumerator _invalidate]: (null)
> debug	12:22:51.621316+0100	FileProviderExt	(504) dropping voucher due to nil personaUniqueString
> debug	12:22:51.621397+0100	FileProviderExt	[DEBUG] ┏2e7 <private>: invalidating vendor enumeration: <private>
> debug	12:22:51.621471+0100	FileProviderExt	Enumerator is being invalidated for item with identifier: 0074594472f651ffb41de
> debug	12:22:51.620773+0100	fileproviderd	[DEBUG] ┣4e9d dispatching to <private>
> debug	12:22:51.621562+0100	FileProviderExt	[DEBUG] ┗2e7
> debug	12:22:51.621632+0100	FileProviderExt	[DEBUG] Attempting to restore persona in -[FPXEnumerator _invalidate]: (null)
> debug	12:22:51.621071+0100	fileproviderd	[DEBUG] ┳4e9d continuing on <private>
> debug	12:22:51.621716+0100	FileProviderExt	(504) dropping voucher due to nil personaUniqueString
> debug	12:22:51.621196+0100	fileproviderd	[DEBUG] ┣4e9f dispatching to <private>
> debug	12:22:51.621830+0100	FileProviderExt	[DEBUG] Attempting to restore persona in -[FPXEnumerator _invalidate]: (null)
> debug	12:22:51.621336+0100	fileproviderd	[DEBUG] ┗4e9d
> debug	12:22:51.622037+0100	FileProviderExt	(504) dropping voucher due to nil personaUniqueString
> debug	12:22:51.621611+0100	fileproviderd	[DEBUG] ┳4e9f continuing on <private>
> debug	12:22:51.622477+0100	FileProviderExt	[DEBUG] Attempting to restore persona in -[FPXEnumerator dealloc]: (null)
> debug	12:22:51.621794+0100	fileproviderd	[DEBUG] ┗4e9f
> debug	12:22:51.622659+0100	FileProviderExt	(504) dropping voucher due to nil personaUniqueString
> debug	12:22:51.622884+0100	FileProviderExt	[DEBUG] <private>: dealloc
> ```

Unfortunately this is of no use, I'd need the logs as per [the release post](https://github.com/nextcloud/desktop/issues/1337#issuecomment-1478386641)

**Comment by claucambra** [comment]
> First, thanks a lot! It's great to to see Nextcloud virtual files for macOS coming! 🙏

❤️

> The beta setup on a Ventura Silicon Mac was fine. Although at first I didn't get I have to check the location 'Nextcloud macOS VFS Beta', not the folder I configured during the setup.. 8)

Folder configuration in the setup wizard is only for the traditional sync method, not for File Provider. This is still a WIP. 🙂

> It took a while to initialise but now the VFS response times are quite ok. Btw, the server is NC25 with notify_push enabled.
> 
> My first question (it won't be the last :), the modification times are set to SkyNet' Big Bang. I've read about this above. However, is it a known issue or a misconfiguration on my side?

Known issue, I can replicate on all my files too. Looking into a fix now.

**Comment by claucambra** [comment]
> First, thanks a lot! It's great to to see Nextcloud virtual files for macOS coming! 🙏

❤️

> The beta setup on a Ventura Silicon Mac was fine. Although at first I didn't get I have to check the location 'Nextcloud macOS VFS Beta', not the folder I configured during the setup.. 8)

Folder configuration in the setup wizard is only for the traditional sync method, not for File Provider. This is still a WIP. 🙂

> It took a while to initialise but now the VFS response times are quite ok. Btw, the server is NC25 with notify_push enabled.
> 
> My first question (it won't be the last :), the modification times are set to SkyNet' Big Bang. I've read about this above. However, is it a known issue or a misconfiguration on my side?

Known issue, I can replicate on all my files too. Looking into a fix now.

**Comment by claucambra** [comment]
Going back to answer some more of these, @marcotrevisan

> Just for the records, some features that I noticed to be missing in Finder:
> 
> * Thumbnails on the icons of unmaterialized files, I see in iCloud they're available (they also show up at space-bar-pressed, not great resolution but enough to recognize the file contents), I guess the API allows for that;

Yup, this is not implemented yet -- on the todo-list

> * When I right-click on a folder containing unmaterialized stuff, the info panel is not showing the total size (a "--" is shown instead).

We generally are not properly handling full folder sizing yet -- also on the todo-list ;)

> About Big Sur: the mount is showing the following issues on write operations:
> 
> * Wen a file / folder is created / updated / deleted, the first level folder containing it will suddenly close and show up as "access denied" (i.e. the folder icon with the red symbol on it). After a second or so, it will become accessible again.

Hmm, strikes me as incorrect error handling in the code, will look into it

>   * If you were creating a subfolder, you'll find it with the default "unnamed folder" name, and when you rename it, the same thing will happen (the rename will be successful though).

AFAIK macOS by default immediately creates a folder with this name -- any name you type is just renaming the already created folder

> Thanks!

Thanks for testing so throughly!

**Comment by melMass** [comment]
Is there a PR to follow progress or something like that?
I've been searching for a while now with answers from 2018 to an hour ago.
Do we still need `showExperimentalOptions=true` ? Are there real issues to be aware of?

Thanks

**Comment by marcotrevisan** [comment]
> Is there a PR to follow progress or something like that?

Up until now: https://github.com/nextcloud/desktop/pull/5527

Cheers

**Comment by marcotrevisan** [comment]
I found another glitch: 
- Added my user to a group;
- The new group membership adds another folder to the visible ones (that folder was shared by another user to that group, read/write enabled);
- The newly discovered folder shows up in FIles (web app) as expected;
- The FileProvider started issuing PROPFINDs on the whole new folder hierarchy;
- After the full tree visit was finished (PROPFINDs finished) the folder didn't show up in Finder (and `fileproviderctl ls` didn't log that folder addition). 
- Only when I made a change via web browser in that folder (deleted an empty sub folder), `fileproviderctl ls` finally logged the incoming change, and Finder finally showed up the new root folder.

So basically looks as if it missed a "trigger" to show up a newly shared folder...

Hope this is reproducible on your systems too.

**Comment by sakura-ikeda** [comment]
Please excuse my poor English. Thanks for the great development.

In the case of LDAP, the URL of Webdav becomes UUID from the user name.
`https://nextcloud.com/remote.php/dav/files/XXXXXXXX-XXXX-XXXXX-XXXX-XXXXXXXXXXXX`
An error occurred in the Finder in this case.
![image](https://user-images.githubusercontent.com/12285089/230309362-80782aed-59ed-4523-a274-db2e64d550c7.png)

**Console app logs**
_The domain name and username have changed._
- debug	16:28:51.952308+0900	FileProviderExt	Starting to read serverUrl: https://nextcloud.com/remote.php/dav/files/username for user: username https://nextcloud.com at depth 1. NCKit info: userId: username, password is empty: NOT EMPTY PASSWORD, urlBase: https://nextcloud.com, ncVersion: 25
- error	16:28:55.307779+0900	FileProviderExt	1 depth readFileOrFolder of url: https://nextcloud.com/remote.php/dav/files/username did not complete successfully, received error: Principal with name username not found
- error	16:28:55.307852+0900	FileProviderExt	Finishing enumeration for user: username https://nextcloud.com with serverUrl: https://nextcloud.com/remote.php/dav/files/username with error Principal with name username not found
- debug	16:28:55.310014+0900	FileProviderExt	Enumerator is being invalidated for item with identifier: NSFileProviderRootContainerItemIdentifier

**Comment by marcotrevisan** [comment]
> I found another glitch:
> 
>     * Added my user to a group;
> 
>     * The new group membership adds another folder to the visible ones (that folder was shared by another user to that group, read/write enabled);
> 
>     * The newly discovered folder shows up in FIles (web app) as expected;
> 
>     * The FileProvider started issuing PROPFINDs on the whole new folder hierarchy;
> 
>     * After the full tree visit was finished (PROPFINDs finished) the folder didn't show up in Finder (and `fileproviderctl ls` didn't log that folder addition).
> 
>     * Only when I made a change via web browser in that folder (deleted an empty sub folder), `fileproviderctl ls` finally logged the incoming change, and Finder finally showed up the new 1st level folder.
> 
> 
> So basically looks as if it missed a "trigger" to show up a newly shared folder...
> 
> Hope this is reproducible on your systems too. Notify_push is enabled in the NC server instance under test.

Sorry I forgot to mention: my macbook was rebooted in the middle of the tree visit process. The process restarted after the reboot (PROPFINDs restarted from where they were interrupted). 

Cheers!

**Comment by d235j** [comment]
Seems to be working well for me, once I set up notify_push and fixed its configuration. Some better error reporting for when notify_push is not set up correctly would be useful.

Separately — can the ability to evict a file/folder be added? At the moment this is not present, which will cause local storage to fill up. Using `fileproviderctl` to attempt to evict seems to also fail with an error. IMO this is probably the most important feature to have after basic functionality.

**Comment by marcotrevisan** [comment]
Hi all. @claucambra found a small file naming glitch and was able to reproduce it. Folders containing "._" in their name get their name truncated once synced (unless there are also spaces in the name, in which case the folder will show up correctly).
For example, "5._TEST" will show up in Finder as "5".

To reproduce it I created the following folders:
- "5._TEST_NO_SPC" (created using the file provider)
- "6._TEST W SPC" (created using the file provider)
- "7._TEST_NO_SPC2" (created via Files web app).

What happened at first is:
- the folder names were correct and, for those created using the provider, they showed up correctly in Files webapp;
- the third folder also showed up in Finder correctly.

But after a while, those folders in Finder were showing up as follows:
- "5"
- "6._TEST W SPC"
- "7" 

Fortunately, their contents are OK. Updates to their content works as if nothing strange has happened.

on the NC server filesystem side they show up as follows: 
```
drwxr-xr-x 2 www-data www-data 4096 Apr 12 11:17  5._TEST_NO_SPC
drwxr-xr-x 2 www-data www-data 4096 Apr 12 11:17 '6._TEST W SPC'
drwxr-xr-x 2 www-data www-data 4096 Apr 12 11:18  7._TEST_NO_SPC2

```

and here is the difference: the one with spaces is deliminted by ' characters which seems to stop the harmful name processing.
Hoping this helps. Best regards!

**Comment by marcotrevisan** [comment]
More info on the above.

1. The same happens if I "touch" a regular file.
2. In a terminal window, 'ls -l' shows the correct file and folder names, so this seems to be a Finder issue somehow.

**Comment by claucambra** [comment]
> Please excuse my poor English. Thanks for the great development.

Your English is great, no apologies needed

> In the case of LDAP/AD, the URL of Webdav becomes UUID from the user name. `https://nextcloud.com/remote.php/dav/files/XXXXXXXX-XXXX-XXXXX-XXXX-XXXXXXXXXXXX` An error occurred in the Finder in this case. ![image](https://user-images.githubusercontent.com/12285089/230309362-80782aed-59ed-4523-a274-db2e64d550c7.png)
> 
> **Console app logs** _The domain name and username have changed._
> 
> * debug	16:28:51.952308+0900	FileProviderExt	Starting to read serverUrl: https://nextcloud.com/remote.php/dav/files/username for user: username https://nextcloud.com at depth 1. NCKit info: userId: username, password is empty: NOT EMPTY PASSWORD, urlBase: https://nextcloud.com, ncVersion: 25
> * error	16:28:55.307779+0900	FileProviderExt	1 depth readFileOrFolder of url: https://nextcloud.com/remote.php/dav/files/username did not complete successfully, received error: Principal with name username not found
> * error	16:28:55.307852+0900	FileProviderExt	Finishing enumeration for user: username https://nextcloud.com with serverUrl: https://nextcloud.com/remote.php/dav/files/username with error Principal with name username not found
> * debug	16:28:55.310014+0900	FileProviderExt	Enumerator is being invalidated for item with identifier: NSFileProviderRootContainerItemIdentifier

Indeed LDAP authentication has not yet been tested, this will require some changes on the desktop client->file provider communication to get working, but thanks for the logs -- it's on the list of things to get right

**Comment by claucambra** [comment]
> I found another glitch:
> 
> * Added my user to a group;
> * The new group membership adds another folder to the visible ones (that folder was shared by another user to that group, read/write enabled);
> * The newly discovered folder shows up in FIles (web app) as expected;
> * The FileProvider started issuing PROPFINDs on the whole new folder hierarchy;
> * After the full tree visit was finished (PROPFINDs finished) the folder didn't show up in Finder (and `fileproviderctl ls` didn't log that folder addition).
> * Only when I made a change via web browser in that folder (deleted an empty sub folder), `fileproviderctl ls` finally logged the incoming change, and Finder finally showed up the new 1st level folder.
> 
> So basically looks as if it missed a "trigger" to show up a newly shared folder...
> 
> Hope this is reproducible on your systems too. Notify_push is enabled in the NC server instance under test.

Do you have any logs of when this initial PROPFIND takes place? Either it's not discovering it on lookup or it's ignoring it for some reason, would be critical to figure out which one is happening and why

**Comment by claucambra** [comment]
> Seems to be working well for me, once I set up notify_push and fixed its configuration. Some better error reporting for when notify_push is not set up correctly would be useful.

Notify_push not being enabled is not really an "error" as the file provider module works fine without push notifications, it's just more efficient with them available.

> Separately — can the ability to evict a file/folder be added? At the moment this is not present, which will cause local storage to fill up. Using `fileproviderctl` to attempt to evict seems to also fail with an error. IMO this is probably the most important feature to have after basic functionality.

Yes, this is on the to-do list

**Comment by wrench12435678** [comment]
Does anyone have a link for the build for this? Can only find an app image on the pull request

(sorry if this is the wrong place to ask, couldn't find it in github actions or a post via the build bot)

**Comment by marcotrevisan** [comment]
> > Sorry I forgot to mention: my macbook was rebooted in the middle of the tree visit process. The process restarted after the reboot (PROPFINDs restarted from where they were interrupted).
> 
> My hunch is that the propfind for the folder in question happened and was recorded in the database, but when the restart happened these changes had not yet been notified through the file provider API -- so when the rescan happened after reboot this folder was not notified as since it was in the database already it was considered unchanged. Will need to look into this, and this is just a hunch, but it makes sense

I did a new test using a small folder so the whole process could complete quickly.
It worked properly. I did it on a group folder, making it appear and disappear by means of both sharing it directly with the current user, unsharing, or adding the user to a group that has visibility on that folder. 
In all cases it worked properly: the folder showed up in Finder when shared and disappeared when unshared.

My issue was therefore caused by the reboot of the system in the middle of the process. I wouldn't underestimate the frequency of this use case, it may occur regularly when large folders are involved in the share/unshare function.
I'm not sure wether a simple "stop" of the system (instead of a reboot) may lead to the same result. If it does, the propbability of that issue arising will be higher...

Thanks and regards!

**Comment by marcotrevisan** [comment]
Hi @claucambra, in your opinion is this a good time for releasing a second beta build?
I saw some stuff and fixes in the commit comments, so I was wondering if it could be worth doing...

Thanks and regards!

**Comment by marcotrevisan** [comment]
Hi @claucambra ,

I've noticed that sometimes, at least under these circumstances:

- Mac OS Monterey
- Big NC mount (many files and folders)

when I navigate into a subfolder it doesn't update its contents. It stays stuck for many seconds, then after a while it shows only partial (seemingly old) content.
The log only shows rows like below: 

```
"Set up enumerator for user: marco.trevisan https://nextcloud.gate39.it with serverUrl: https://nextcloud.<...>"
"Providing enumerator for item with identifier: <...>"
"Enumerator is being invalidated for item with identifier: <...>"
"Received item request for item with identifier: <...>"
```

Output of `fileproviderctl ls <...>` for that subfolder:

`
 <my_folder> id:__fp/fs/fileID(70701226) ⬇  (indeterminate)  ieri, 17:43 (used)
`

where all the other subfolders in that directory are not `(indeterminate)`.

I tried to add and remove files/folders under that subfolder, and the push notifications arrive and trigger PROPFINDs as usual, the updates are also logged in `fileproviderctl`, but they're not having the desired effect.

Is there a way to tell it to "resync" a sub folder?

Thanks and regards

**Comment by misku** [comment]
@claucambra Thank you for the great work! 👍

**Comment by marcotrevisan** [comment]
Thanks @claucambra and Nextcloud !!!

**Comment by Nusephived** [comment]
@misku @marcotrevisan @helmut72 @claucambra Thanks guys for this new amazing feature! Is there a guide to use the virtual file system? I don't find it in settings.

**Comment by melMass** [comment]
Awesome, which client version has it?

**Comment by jesusdeveloper** [comment]
> Awesome, which client version has it?

I want know this also :)

**Comment by marcotrevisan** [comment]
> @misku @marcotrevisan @helmut72 @claucambra Thanks guys for this new amazing feature! Is there a guide to use the virtual file system? I don't find it in settings.

Lol.. Don’t credit me for this, I only did a bit of beta testing :)
For what I’ve seen it’s been merged and it MAY be added to the 3.9.0 milestone (you might want to watch for it under the Issues section). I don’t know if it will actually land in 3.9.x.

The beta build used here has no other documentation than this: https://github.com/nextcloud/desktop/issues/1337#issuecomment-1478386641

**Comment by melMass** [comment]
> Lol.. Don’t credit me for this, I only did a bit of beta testing :) 

That was still quite useful and helped to complete this PR so they are right: Thanks everyone involved :)

**Comment by helmut72** [comment]
> @misku @marcotrevisan @helmut72 @claucambra Thanks guys for this new amazing feature! Is there a guide to use the virtual file system? I don't find it in settings.

Also no credit for me. I just opened the feature request. ;)

**Comment by misku** [comment]
> @misku @marcotrevisan @helmut72 @claucambra Thanks guys for this new amazing feature! Is there a guide to use the virtual file system? I don't find it in settings.

I just thanked the original developer for the great work ;-D

**Comment by mamboone** [comment]
A personal thanks to all of you from my side as well. Great job!!!
Love to see this feature finally arriving within Nextcloud which I looked for eagerly after being used to it within OneDrive Business for a long time.
I just upgraded my NAS to the latest 25 version yesterday (25.6) (I cannot update to 26 via the integrated updater - don't know why. 26 will just not be offered to me).
Therefore the same question as Nusephived had it above: any documentation how to use it - even in beta state? And what the requirements are?
I have Mac OS Nextcloud Desktop 3.8.1 and experimental features enabled.
But it just shows me the "suffix" based virtual files handling which I do NOT want to use.
Which server version is needed? Which client version? Which settings?
Thanks a lot!

**Comment by claucambra** [comment]
Hey all, no need to give thanks, I wanted to use this as much as you did 🙂

Now that the relevant PR has been merged, the file provider module will be included in version 3.9. The module is now also already included in the unstable daily builds of the desktop client. 

For the moment, the file provider module is hidden and disabled by default as we have some rough edges still to iron out. It can, however, be enabled by adding `macFileProviderModuleEnabled = true` to your client's `nextcloud.cfg` file. **Be warned that this module is still not ready for use in production environments!**

Finally, we are also providing a new version of the macOS VFS Beta version of the desktop client that can be installed alongside the normal desktop client. You can download it here:

https://cloud.nextcloud.com/s/8JnmYA8a7xDm9tT

Regarding the issues that you guys have brought up:

1. @marcotrevisan Unfortunately I haven't yet had the time to properly resolve several of the issues you were facing with borked change fetching after system shutdown, or the other issues. This is on the todo list for now and will need to be resolved before considering this feature stable
2. Initial sync is still very slow and this will need to also be resolved
3. There is no GUI management of the file provider sync module, this is -- you guessed it -- also on the todo list.

**Comment by marcotrevisan** [comment]
Hi @claucambra, thanks for your feedback!

As for the open points: since this ticket is closed and is full of comments, if you prefer I may create separate tickets -one for each open point- that link to this.
I'll download the new beta later today and will be able to check for possible resolution of some of the open points.

Best regards!

**Comment by marcotrevisan** [comment]
Hi @BJKle,
my 2 cents on your questions:

> a) Is there a minimum server version required? b) Do we need modules or a special configuration on the backen?

a) NC server v25.x is quite sufficient. Lower versions should work, if I remember correctly @claucambra wrote he developed it against the NC 23 webdav API. 
b) Notify_push is suggested but not mandatory. It should work with a "defaults" configured server. For the sake of testing I would recommend against external storage folders, but that is to keep things easier to reproduce. 

Hope this helps. Cheers!

**Comment by Saifudaullah** [comment]
> Hey all, no need to give thanks, I wanted to use this as much as you did 🙂
> 
> Now that the relevant PR has been merged, the file provider module will be included in version 3.9. The module is now also already included in the unstable daily builds of the desktop client.
> 
> For the moment, the file provider module is hidden and disabled by default as we have some rough edges still to iron out. It can, however, be enabled by adding `macFileProviderModuleEnabled = true` to your client's `nextcloud.cfg` file. **Be warned that this module is still not ready for use in production environments!**
> 
> Finally, we are also providing a new version of the macOS VFS Beta version of the desktop client that can be installed alongside the normal desktop client. You can download it here:
> 
> https://cloud.nextcloud.com/s/8JnmYA8a7xDm9tT
> 
> Regarding the issues that you guys have brought up:
> 
> 1. @marcotrevisan Unfortunately I haven't yet had the time to properly resolve several of the issues you were facing with borked change fetching after system shutdown, or the other issues. This is on the todo list for now and will need to be resolved before considering this feature stable
> 2. Initial sync is still very slow and this will need to also be resolved
> 3. There is no GUI management of the file provider sync module, this is -- you guessed it -- also on the todo list.

Hi @claucambra, I'm unable to access the provided link. Can you provide another link for me to test this beta client?

**Comment by BJKle** [comment]
@claucambra Any news or timeline?

**Comment by marcotrevisan** [comment]
I'm using the beta 2 client for a while now, under Mac OS Ventura.
I'd like to encourage the team to "do the last mile" asap, because even if it has its own glitches, it's also already showing some great advantages over the existing implememtation (not only the NC Client, but other webdav clients also).
Thanks!

**Comment by kklem0** [comment]
Can I test it with the daily build since it is still in the master branch?
https://download.nextcloud.com/desktop/daily/

**Comment by claucambra** [comment]
> Can I test it with the daily build since it is still in the master branch? https://download.nextcloud.com/desktop/daily/

We had a major regression with the Finder integration caused by the File Provider module so we had to disable building of the File Provider module by default. So the dailies do not have the module enabled. If you are brave enough to try building the client yourself, however, you can toggle the compilation flag and build the client with the module enabled

**Comment by pete1019** [comment]
Hi, is the M1 support with virtual drive back in the daily?
Otherwise where to get one without need to build?
Thanks

**Comment by marcotrevisan** [comment]
I'm just guessing but I'd expect to see this being closed first: https://github.com/nextcloud/desktop/issues/5693
You might want to follow the activity on this feature branch: https://github.com/nextcloud/desktop/tree/feature/file-provider-config-ui

My 2 cents

**Comment by szaimen** [comment]
Reopening because fileprovider support was reverted iirc

**Comment by BJKle** [comment]
Is there a target release?

**Comment by claucambra** [comment]
> Reopening because fileprovider support was reverted for the time being iirc

Reclosing, we are shipping working code in master at the moment ( we just are not enabling by default in normal releases as we are in the process of creating a specific version of the client with the file provider engine )

We also have the VFS betas for use at the moment

> Is there a target release?

At the moment no, we are being cautious about deadlines considering we have had unexpected hiccups when it comes it existing integrations, for instance. At the moment we are working on getting the FIle Provider engine better integrated into our GUI and on improving initial sync times; once this is done, we will probably be ready to release something we are happy to call stable :)

**Comment by Saifudaullah** [comment]
@claucambra How do we test the beta? Is there any link that I can download the beta?

**Comment by marcotrevisan** [comment]
> @claucambra How do we test the beta? Is there any link that I can download the beta?

I see the beta package here (not sure if it's the latest version of it): 
(folder) https://download.nextcloud.com/desktop/prereleases/virtualdrive/
https://download.nextcloud.com/desktop/prereleases/virtualdrive/Nextcloud-VFS-beta-macOS.pkg

**Comment by Saifudaullah** [comment]
@marcotrevisan I'm afraid that is not the latest one. The timestamp of the installer is 2023-03-21.

**Comment by claucambra** [comment]
Hi all, we've built a new version of the File Provider based VFS beta based on the current stable branch of the desktop client.

You may download the macOS VFS Beta 3 from the link below:

https://download.nextcloud.com/desktop/prereleases/virtualdrive/Nextcloud-stable-3.10-fileprovider.pkg

Key changes in this beta include some small stability improvements, as well as all changes in the latest 3.10 desktop client :)

**Comment by marcotrevisan** [comment]
Hi @claucambra, thanks very much for the new beta!
I noticed that now hitting the space bar on a not-already-materialized item will cause materialization and subsequent preview, nice!
I still can't get precisely under what conditions (maybe filename patterns?), but looks like the previews that available on the server aren't actually used for the icons of unmaterialized files, except for some of them in my private folders. 
That's weird... is it a known issue?

Thanks!

**Comment by Saifudaullah** [comment]
Thank you so much for the new beta @claucambra. I've installed it, and I assume to enable the file provider, we need to add macFileProviderModuleEnabled = true to the client's nextcloud.cfg.

I can say that this beta seems promising and quite nice to be used. Now it really feels like a proper implementation of virtual file system. 

Below is my feedback:
1. The initial sync is really slow.
2. The loading indication spinning forever when I open empty folder
3. Folders not showing size
4. Folders that are already synced (downloaded) still showing dotted cloud icon

I also has a question regarding Nextcloud folder.  Is it normal that we have two Nextcloud folders? One under "/Users/<username>/Nextcloud" and another one under "/Users/<username>/Library/CloudStorage/Nextcloud-<nextcloud_username>@<domain>/"?

**Comment by marcotrevisan** [comment]
@Saifudaullah I didn't have to set macFileProviderModuleEnabled = true in my config.

This build is still including both sync engines, so I'd suggest you to uncheck the root folder to practically disable the legacy engine. The legacy engine is responsible for "/Users/"username"/Nextcloud" (it will still be there but it won't sync).

As you noticed the initial sync is slow and until it's finished you will experience delays in the UI. 
Your point 2 is not showing up in my client so I'd suggest to wait a bit more before starting your test.
I can also confirm your points 3. and 4. Point 3 is a known issue and in to-do list if I remember correctly.

Regards,

**Comment by claucambra** [comment]
> Thank you so much for the new beta @claucambra. I've installed it, and I assume to enable the file provider, we need to add macFileProviderModuleEnabled = true to the client's nextcloud.cfg.
> 
> I can say that this beta seems promising and quite nice to be used. Now it really feels like a proper implementation of virtual file system.
> 
> Below are my feedback:
> 
>     1. The initial sync is really slow.
> 
>     2. The loading indication spinning forever when I open empty folder

Points 1 and 2 are related. At the moment we require a full scan of the server folder hierarchy during initial sync. We are considering how to address this

>     3. Folders not showing size

Yeah, aware of this bug -- thanks for raising

>     4. Folders that are already synced (downloaded) still showing dotted cloud icon

Can reproduce, still not 100% sure why this is the case. Will investigate

**Comment by claucambra** [comment]
As an aside, I've opened a Discussions thread to continue any discussion on the VFS beta as discussing on a closed thread is not ideal :)

https://github.com/nextcloud/desktop/discussions/6267

**Comment by claucambra** [comment]
> Do you think the problematic issue you mentioned earlier is fixed?

I can confirm it has been fixed, but there is unfortunately no way we are going to be able to ship both the Finder Sync extension and the File Provider module alongside each other. 3.11 is going to be released imminently and I can confirm we will not be including the File Provider module with this release either.

When this module is considered stable we will be shipping a new client edition alongside the normal client which foregoes the Finder integration for the conventional sync engine in favour of the File Provider module. It is likely we will also strip out the traditional sync engine from the VFS-focused edition of the desktop client.

None of this is final and is subject to change, though! Thoughts welcome.

**Comment by marcotrevisan** [comment]
forgot to mention: that glich is happening to me since the first beta.
If I remember correctly (not 100% sure) beta2 made it partially work.

By the way, I noticed in my system (M1-Sonoma) the command “fileprovoderctl ls _domainname_” seems not be implemented anymore, but the man page is unvhanged… ??
Oany clues in new ways to accomplishing that are very welcome!
Thanks!

>I still can't get precisely under what conditions (maybe filename patterns?), but looks like the previews that available on the server aren't actually used for the icons of unmaterialized files, except for some of them in my private folders. That's weird... is it a known issue?

**Comment by funkemunky** [comment]
When I download a file pressing "Download Now", the file does indeed download. However, the "Remove download" button is missing when I want to set it to cloud-only.
